src/pc/1.17/ChunkColumn.js

```diff
@@ -0,0 +1,275 @@
+    }
+
+    getBiomeIndex (pos) {
+      const i = (pos.x >> 2) & 3
+      const j = (pos.y >> 2) - (this.minWorldHeight >> 2)
+      const k = (pos.z >> 2) & 3
+      return j << 4 | k << 2 | i
+    }
+
+    getBottomSectionCoord () {
+      return getSectionCoord(this.minWorldHeight)
+    }
+
+    getTopSectionCoord () {
+      return getSectionCoord(this.maxWorldHeight - 1) + 1
+    }
+
+    isBlockPosValid (pos) {
+      return pos.y >= this.minWorldHeight && pos.y < this.maxWorldHeight &&
+        pos.x >= 0 && pos.x < constants.SECTION_WIDTH &&
+        pos.z >= 0 && pos.z < constants.SECTION_WIDTH
+    }
+
+    // Block State retrieval methods
+    getBlockStateId (pos) {
+      if (this.isBlockPosValid(pos)) {
+        const section = this.sections[this.getSectionIndex(pos.y)]
+        return section ? section.getBlock(toSectionPos(pos)) : 0
+      }
+      return 0
+    }
+
+    setBlockStateId (pos, stateId) {
+      if (!this.isBlockPosValid(pos)) {
+        return
+      }
+
+      const sectionIndex = this.getSectionIndex(pos.y)
+      let section = this.sections[sectionIndex]
+
+      if (!section) {
+        // if it's air, do not create a new section for it
+        if (stateId === 0) {
+          return
+        }
+        section = new ChunkSection()
+        this.sectionMask |= 1 << sectionIndex
+        this.sections[sectionIndex] = section
+      }
+      section.setBlock(toSectionPos(pos), stateId)
+    }
+
+    // Biome methods
+    getBiome (pos) {
+      if (this.isBlockPosValid(pos)) {
+        return this.biomes[this.getBiomeIndex(pos)]
+      }
+      return 0
+    }
+
+    setBiome (pos, biome) {
+      if (this.isBlockPosValid(pos)) {
+        this.biomes[this.getBiomeIndex(pos)] = biome
+      }
+    }
+
+    // Convenience methods returning Block object
+    getBlock (pos) {
+      const biome = this.getBiome(pos)
+      const stateId = this.getBlockStateId(pos)
+
+      return Block.fromStateId(stateId, biome)
+    }
+
+    setBlock (pos, block) {
+      if (typeof block.stateId !== 'undefined') {
+        this.setBlockStateId(pos, block.stateId)
+      }
+      if (typeof block.biome !== 'undefined') {
+        this.setBiome(pos, block.biome.id)
+      }
+    }
+
+    // Legacy conversion methods accepting block type + metadata
+    getBlockType (pos) {
+      const blockStateId = this.getBlockStateId(pos)
+      return mcData.blocksByStateId[blockStateId].id
+    }
+
+    getBlockData (pos) {
+      const blockStateId = this.getBlockStateId(pos)
+      return mcData.blocksByStateId[blockStateId].metadata
+    }
+
+    setBlockType (pos, id) {
+      this.setBlockStateId(pos, mcData.blocks[id].minStateId)
+    }
+
+    setBlockData (pos, data) {
+      this.setBlockStateId(pos, mcData.blocksByStateId[this.getBlockStateId(pos)].minStateId + data)
+    }
+
+    // Skylight data has been removed in 1.17, now it is computed client-side automatically
+    // and never synced from the server in chunk packets
```

Comment:
The light update packet still exists https://wiki.vg/Protocol#Update_Light so this claim should be double-checked.

Reply:
fair point, brought lightning implementation back as requested before